### DISCUSSION TITLE
useFieldArray: add multiple records at once

### DISCUSSION THREAD

**Discussion by solace** [react-hook-form/react-hook-form#3105] [Q&A] [answered]
Apologies if I've missed a similar post, but I have need of adding multiple records to a nested field array (csv upload) and I'm uncertain of the best way to approach this.

1. Iterating `append` is inefficient for large numbers of inserts.
2. `reset` to update the entire form object seems a bit overkill, and for some reason gives a `Can't perform a React state update on an unmounted component` which seems to result from RHF's `setFields` function internally. I am currently unable to repro with the nestedFieldArrays demo but it seems to have something to do with RHF updates causing components to unmount. I'm not sure if this is related to the fact that I am sharing ref usage.
3. `setValue` and `watch` works great, but this means I cannot use `append` or `remove` to add and remove individual items because `fields` and `watch` do not sync.

What is the recommended method for insertion of multiple items at once into a field array?

Also, I see in https://spectrum.chat/react-hook-form/help/usefieldarray-doesnt-update-fields~38f3406b-6237-410e-a357-940f6895f957?authed=true that you have suggested merging watch and fields, but do you have a recommended way of doing this?

Thank you.

**Comment by bluebill1049** [comment, marked as answer]
`append([])`?

**Reply by solace** [reply]
Yes, thank you.

I clearly over-thought how this worked.

**Reply by segullshairbutt** [reply]
Thank you so much. Sometimes little things are more large.

**Reply by normandanielmq** [reply]
Awesome, how can we do the opposite? Delete all records att once?

**Reply by bluebill1049** [reply]
`remove()` it's in the doc.

**Reply by segullshairbutt** [reply]
> 
> 
> Awesome, how can we do the opposite? Delete all records att once?

You can append an empty array.

**Reply by eddysapata** [reply]
Bill You are Awesome :) <3 Love you :)

**Reply by bluebill1049** [reply]
> Bill You are Awesome :) <3 Love you :)

You too ❤️ thanks

**Reply by Johnrobmiller** [reply]
Not sure this is the best dx. Devs who want to append multiple items and don't know about this special use of `append` are probably just going to stack multiple calls to `append` one on top of the other, like so:

```
append(...) // first item
append(...) // second item
append(...) // etc.
```

When that fails, then waste valuable time troubleshooting, logging, and resort to google. The best dx is the one that is most intuitive, the one devs are going to try the first time before resorting to googling it.

So, the best way to have coded this would have been that multiple calls to `append` work as expected with no special instructions required.
